package com.odilo.techchallenge.repository;

import com.odilo.techchallenge.entity.BookEntity;
import jakarta.persistence.LockModeType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;

import java.util.Optional;

public interface BookRepository extends JpaRepository<BookEntity, String> {

    Page<BookEntity> findByTitleContainingIgnoreCase(String title, Pageable pageable);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    Optional<BookEntity> findWithLockByIsbn(String isbn);
}
